system/xmllib/libxml_xpathprocessor.cpp

```diff
@@ -149,11 +539,20 @@ class CLibXpathContext : public CInterface, public IXpathContext
         return evaluateAsString(evaluate(clibCompiledXpath->getCompiledXPathExpression(), compiledXpath->getXpath()), evaluated, compiledXpath->getXpath());
     }
 
+    virtual double evaluateAsNumber(ICompiledXpath * compiledXpath) override
+    {
+        CLibCompiledXpath * clibCompiledXpath = static_cast<CLibCompiledXpath *>(compiledXpath);
+        if (!clibCompiledXpath)
+            throw MakeStringException(XPATHERR_MissingInput,"XpathProcessor:evaluateAsNumber: Error: Missing compiled XPATH");
+        return evaluateAsNumber(evaluate(clibCompiledXpath->getCompiledXPathExpression(), compiledXpath->getXpath()), compiledXpath->getXpath());
+    }
+
 private:
     virtual bool setXmlDoc(const char * xmldoc) override
     {
         if (xmldoc && * xmldoc)
         {
+            WriteLockBlock rblock(m_rwlock);
```

Comment:
Is it better to move xmlParseDoc() out of the lock block (by assigning to a local variable first)?